internal/graph/dotgraph.go

```diff
@@ -483,9 +483,16 @@ func escapeAllForDot(in []string) []string {
 	return out
 }
 
-// escapeForDot escapes double quotes and backslashes, and replaces Graphviz's
-// "center" character (\n) with a left-justified character.
+// escapeForDot escapes double quotes and backslashes, and replaces newlines
+// with a left-justified escape (\l).
 // See https://graphviz.org/docs/attr-types/escString/ for more info.
 func escapeForDot(str string) string {
 	return strings.ReplaceAll(strings.ReplaceAll(strings.ReplaceAll(str, `\`, `\\`), `"`, `\"`), "\n", `\l`)
 }
+
+// escapeForDotCentered escapes double quotes and backslashes, and replaces
+// newlines with Graphviz's center escape (\n).
+// See https://graphviz.org/docs/attr-types/escString/ for more info.
+func escapeForDotCentered(str string) string {
```

Comment:
I might be missing something but quoting myself from https://github.com/google/pprof/pull/683:

> I'd also expect any non-printable characters like newlines to be rendered as their C escaped version, so that the tag value is always one line when rendered. Same for function names. The escaping was initially added in https://github.com/google/pprof/pull/557 for graph legend labels that come from the profile comments field and for that we did support newlines to render newlines from comments as newlines since some comments intentionally use that kind of formatting. I think for function names and tag values it should be different and we shouldn't allow any formatting sequences in those strings.

So for function names and tag values I'd expect the newline to be escaped into a printable character, not rendered as either (left or center) actual newline character...

Reply:
The newline handling is confusing, and there could be a better place to put it. However, I believe this version produces the desired output. Here is an example with my Go test output that currently fails to render with pprof, which shows the newlines in the key/value pairs being displayed as `\n`:

<img width="469" alt="Screen Shot 2022-03-22 at 9 45 01 AM" src="https://user-images.githubusercontent.com/675151/159495640-5c802d1a-c0c0-4323-9feb-78a672c93c31.png">

I tried to change the existing logic as little as possible. My logic was the following:

The `joinLabels` function combines multiple key/value pairs, and returns a string as it should be displayed to a human, without any dot/graphviz details. It replaces newline characters with a `\n` sequence, then combines the multiple key/value pairs with newlines, since we want to display multiple key/value pairs on separate lines. The `TestJoinLabels` attempts to demonstrate this.

Next, `escapeForDotCentered` must convert this "human string" into a version that can be embedded into Graphviz's "escString". To do that, it must escape backslashes, escape quotes, and replace newline characters with `\n`. It turns out: I can remove the newline escaping from this function, since "centered text" in graphviz's default. However, this will make this function *different* from `escapeForDot`. Is that desirable?

I totally forgot about other non-printable characters though, so thanks! I will add that to joinLabels. My logic is if we were to us a different graph rendering tool, we would still want this escaping, so it should not be "dot specific".

Reply:
I would suggest we have a single `escapeForDot()` function with a bool argument like `escapeNewlines` and set that argument to true when escaping comments and to false when escaping label values.

The `joinLabels` function should only join lines with newlines (actual ones), it should not do any escaping to the joined values itself, it will assume the values have been escaped as necessary already.